type checking failing due to a dependency update

**Describe the bug**
Clone repo, run `hatch shell`, and then `hatch run test:types`. See below for output
**Error message**
```
Installing collected packages: types-tabulate, types-six, types-simplejson, types-requests, types-PyYAML, types-pycurl, types-Deprecated, types-decorator, types-dateparser
Successfully installed types-Deprecated-1.2.15.20241117 types-PyYAML-6.0.12.20241230 types-dateparser-1.2.0.20240420 types-decorator-5.1.8.20250121 types-pycurl-7.45.4.20241216 types-requests-2.32.0.20241016 types-simplejson-3.19.0.20241221 types-six-1.17.0.20241205 types-tabulate-0.9.0.20241207

haystack/components/generators/hugging_face_api.py:215: error: Argument 1 to "_build_non_streaming_response" of "HuggingFaceAPIGenerator" has incompatible type "TextGenerationOutput | Iterable[TextGenerationStreamOutput]"; expected "TextGenerationOutput"  [arg-type]
haystack/components/generators/hugging_face_api.py:224: error: Incompatible types in assignment (expression has type "TextGenerationStreamOutputToken", variable has type "TextGenerationOutputToken")  [assignment]
haystack/components/generators/chat/hugging_face_local.py:146: error: Incompatible types in assignment (expression has type "str | None", variable has type "Literal['text-generation', 'text2text-generation'] | None")  [assignment]
haystack/components/generators/chat/hugging_face_api.py:260: error: Argument 3 to "_run_non_streaming" of "HuggingFaceAPIChatGenerator" has incompatible type "list[dict[str, Collection[str]]] | None"; expected "list[ChatCompletionInputTool] | None"  [arg-type]
haystack/components/generators/chat/hugging_face_api.py:294: error: Argument 1 to "StreamingChunk" has incompatible type "str | None"; expected "str"  [arg-type]
haystack/components/generators/chat/hugging_face_api.py:299: error: Dict entry 0 has incompatible type "str": "str | None"; expected "str": "str"  [dict-item]
haystack/components/generators/chat/hugging_face_api.py:300: error: Dict entry 1 has incompatible type "str": "str | None"; expected "str": "str"  [dict-item]
haystack/components/generators/chat/hugging_face_api.py:301: error: Dict entry 2 has incompatible type "str": "int"; expected "str": "str"  [dict-item]
haystack/components/generators/chat/hugging_face_api.py:302: error: Dict entry 3 has incompatible type "str": "dict[str, int]"; expected "str": "str"  [dict-item]
haystack/components/generators/chat/hugging_face_api.py:303: error: Dict entry 4 has incompatible type "str": "str | None"; expected "str": "str"  [dict-item]
haystack/components/generators/chat/hugging_face_api.py:347: error: Incompatible types in assignment (expression has type "dict[str, int]", target has type "int | str | None")  [assignment]
```

**Expected behavior**
We should either pin whatever dependency is causing this issue or fix the errors.

**Additional context**
This was working fine before I cleared my UV cache. Now that everything was downloaded fresh, it seems like an update to a package is causing these errors.
**To Reproduce**
`git clone https://github.com/deepset-ai/haystack`
`hatch shell`
`hatch run test:types`

**FAQ Check**
- [x] Have you had a look at [our new FAQ page](https://docs.haystack.deepset.ai/docs/faq)?

**System:**
 - OS: Ubuntu
 - GPU/CPU: N/A
 - Haystack version (commit or version number): d2348ad
 - DocumentStore: N/A
 - Reader: N/A
 - Retriever: N/A
 - Python Version: 3.11

--- a/haystack/components/generators/chat/hugging_face_api.py
+++ b/haystack/components/generators/chat/hugging_face_api.py
@@ -2,6 +2,7 @@
 #
 # SPDX-License-Identifier: Apache-2.0
 
+from copy import copy
 from datetime import datetime
 from typing import Any, Callable, Dict, Iterable, List, Optional, Union
 
@@ -15,6 +16,7 @@ from haystack.utils.url_validation import is_valid_http_url
 
 with LazyImport(message="Run 'pip install \"huggingface_hub[inference]>=0.27.0\"'") as huggingface_hub_import:
     from huggingface_hub import (
+        ChatCompletionInputFunctionDefinition,
         ChatCompletionInputTool,
         ChatCompletionOutput,
         ChatCompletionStreamOutput,
@@ -255,8 +257,15 @@ class HuggingFaceAPIChatGenerator:
 
         hf_tools = None
         if tools:
-            hf_tools = [{"type": "function", "function": {**t.tool_spec}} for t in tools]
-
+            hf_tools = [
+                ChatCompletionInputTool(
+                    function=ChatCompletionInputFunctionDefinition(
+                        name=tool.name, description=tool.description, arguments=tool.parameters
+                    ),
+                    type="function",
+                )
+                for tool in tools
+            ]
         return self._run_non_streaming(formatted_messages, generation_kwargs, hf_tools)
 
     def _run_streaming(
@@ -278,13 +287,12 @@ class HuggingFaceAPIChatGenerator:
             # see https://huggingface.co/docs/huggingface_hub/package_reference/inference_client#huggingface_hub.InferenceClient.chat_completion.n
             choice = chunk.choices[0]
 
-            text = choice.delta.content
-            if text:
-                generated_text += text
+            text = choice.delta.content or ""
+            generated_text += text
 
             finish_reason = choice.finish_reason
 
-            meta = {}
+            meta: Dict[str, Any] = {}
             if finish_reason:
                 meta["finish_reason"] = finish_reason
 
@@ -336,7 +344,11 @@ class HuggingFaceAPIChatGenerator:
                 )
                 tool_calls.append(tool_call)
 
-        meta = {"model": self._client.model, "finish_reason": choice.finish_reason, "index": choice.index}
+        meta: Dict[str, Any] = {
+            "model": self._client.model,
+            "finish_reason": choice.finish_reason,
+            "index": choice.index,
+        }
 
         usage = {"prompt_tokens": 0, "completion_tokens": 0}
         if api_chat_output.usage:

--- a/haystack/components/generators/chat/hugging_face_api.py
+++ b/haystack/components/generators/chat/hugging_face_api.py
@@ -2,7 +2,6 @@
 #
 # SPDX-License-Identifier: Apache-2.0
 
-from copy import copy
 from datetime import datetime
 from typing import Any, Callable, Dict, Iterable, List, Optional, Union
 

--- a/haystack/components/generators/chat/hugging_face_local.py
+++ b/haystack/components/generators/chat/hugging_face_local.py
@@ -145,7 +145,7 @@ class HuggingFaceLocalChatGenerator:
             elif isinstance(huggingface_pipeline_kwargs["model"], str):
                 task = model_info(
                     huggingface_pipeline_kwargs["model"], token=huggingface_pipeline_kwargs["token"]
-                ).pipeline_tag
+                ).pipeline_tag  # type: ignore[assignment]  # we'll check below if task is in supported tasks
 
         if task not in PIPELINE_SUPPORTED_TASKS:
             raise ValueError(

--- a/haystack/components/generators/hugging_face_api.py
+++ b/haystack/components/generators/hugging_face_api.py
@@ -212,6 +212,8 @@ class HuggingFaceAPIGenerator:
         if streaming_callback is not None:
             return self._stream_and_build_response(hf_output, streaming_callback)
 
+        print(type(hf_output))
+
         assert isinstance(hf_output, TextGenerationOutput)  # The output is non streaming, but mypy doesn't know that.
         return self._build_non_streaming_response(hf_output)
 

--- a/haystack/components/generators/hugging_face_api.py
+++ b/haystack/components/generators/hugging_face_api.py
@@ -212,8 +212,6 @@ class HuggingFaceAPIGenerator:
         if streaming_callback is not None:
             return self._stream_and_build_response(hf_output, streaming_callback)
 
-        print(type(hf_output))
-
         assert isinstance(hf_output, TextGenerationOutput)  # The output is non streaming, but mypy doesn't know that.
         return self._build_non_streaming_response(hf_output)
 

--- a/haystack/components/generators/hugging_face_api.py
+++ b/haystack/components/generators/hugging_face_api.py
@@ -4,7 +4,7 @@
 
 from dataclasses import asdict
 from datetime import datetime
-from typing import Any, Callable, Dict, Iterable, List, Optional, Union
+from typing import Any, Callable, Dict, Iterable, List, Optional, Union, cast
 
 from haystack import component, default_from_dict, default_to_dict, logging
 from haystack.dataclasses import StreamingChunk
@@ -212,8 +212,8 @@ class HuggingFaceAPIGenerator:
         if streaming_callback is not None:
             return self._stream_and_build_response(hf_output, streaming_callback)
 
-        assert isinstance(hf_output, TextGenerationOutput)  # The output is non streaming, but mypy doesn't know that.
-        return self._build_non_streaming_response(hf_output)
+        # mypy doesn't know that hf_output is a TextGenerationOutput, so we cast it
+        return self._build_non_streaming_response(cast(TextGenerationOutput, hf_output))
 
     def _stream_and_build_response(
         self, hf_output: Iterable["TextGenerationStreamOutput"], streaming_callback: Callable[[StreamingChunk], None]

--- a/haystack/components/generators/hugging_face_api.py
+++ b/haystack/components/generators/hugging_face_api.py
@@ -17,8 +17,8 @@ with LazyImport(message="Run 'pip install \"huggingface_hub>=0.27.0\"'") as hugg
     from huggingface_hub import (
         InferenceClient,
         TextGenerationOutput,
-        TextGenerationOutputToken,
         TextGenerationStreamOutput,
+        TextGenerationStreamOutputToken,
     )
 
 
@@ -212,6 +212,7 @@ class HuggingFaceAPIGenerator:
         if streaming_callback is not None:
             return self._stream_and_build_response(hf_output, streaming_callback)
 
+        assert isinstance(hf_output, TextGenerationOutput)  # to make mypy happy. We know that this is non streaming.
         return self._build_non_streaming_response(hf_output)
 
     def _stream_and_build_response(
@@ -221,7 +222,7 @@ class HuggingFaceAPIGenerator:
         first_chunk_time = None
 
         for chunk in hf_output:
-            token: TextGenerationOutputToken = chunk.token
+            token: TextGenerationStreamOutputToken = chunk.token
             if token.special:
                 continue
 

--- a/haystack/components/generators/hugging_face_api.py
+++ b/haystack/components/generators/hugging_face_api.py
@@ -212,7 +212,7 @@ class HuggingFaceAPIGenerator:
         if streaming_callback is not None:
             return self._stream_and_build_response(hf_output, streaming_callback)
 
-        assert isinstance(hf_output, TextGenerationOutput)  # to make mypy happy. We know that this is non streaming.
+        assert isinstance(hf_output, TextGenerationOutput)  # The output is non streaming, but mypy doesn't know that.
         return self._build_non_streaming_response(hf_output)
 
     def _stream_and_build_response(

--- a/test/components/generators/test_hugging_face_api.py
+++ b/test/components/generators/test_hugging_face_api.py
@@ -7,6 +7,7 @@ from datetime import datetime
 
 import pytest
 from huggingface_hub import (
+    TextGenerationOutput,
     TextGenerationOutputToken,
     TextGenerationStreamOutput,
     TextGenerationStreamOutputStreamDetails,
@@ -30,7 +31,7 @@ def mock_check_valid_model():
 @pytest.fixture
 def mock_text_generation():
     with patch("huggingface_hub.InferenceClient.text_generation", autospec=True) as mock_text_generation:
-        mock_response = Mock()
+        mock_response = Mock(spec=TextGenerationOutput)
         mock_response.generated_text = "I'm fine, thanks."
         details = Mock()
         details.finish_reason = MagicMock(field1="value")

